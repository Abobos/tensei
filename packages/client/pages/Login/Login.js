import React from 'react'
import { mustBeNotAuthenticated } from '~/store/auth'
import {
    TextField,
    Button,
    CheckboxField,
} from '@contentful/forma-36-react-components'

class LoginPage extends React.Component {
    state = {
        email: '',
        errors: {},
        password: '',
        isLoading: false,
        rememberMe: false,
    }

    login = (submitEvent) => {
        submitEvent.preventDefault()

        this.setState({
            isLoading: true,
        })

        Flamingo.request
            .post('login', {
                email: this.state.email,
                password: this.state.password,
                rememberMe: this.state.rememberMe,
            })
            .then(() => {
                window.location.href = Flamingo.getPath('')
            })
            .catch((error) => {
                if (error?.response?.status === 422) {
                    this.setState({
                        isLoading: false,
                        errors: this.flattenErrors(error.response.data.errors),
                    })
                }
            })
    }

    flattenErrors = (errors) => {
        const flatErrors = {}

        errors.forEach((error) => {
            flatErrors[error.field] = error.message
        })

        return flatErrors
    }

    render() {
        return (
            <div className="w-full bg-gray-100 h-screen">
                <div className="max-w-md mx-auto pt-20 md:px-0 px-5">
                    <div className="flex justify-center mb-5">
                        <img
                            src="https://strapi.katifrantz.com/admin/3f6f46544e110a51499353fdc9d12bfe.png"
                            className="h-10"
                            alt="company-logo"
                        />
                    </div>
                    <div className="border-t-2 border-blue-primary bg-white shadow-md py-8 px-8">
                        <form onSubmit={this.login} action="">
                            <TextField
                                labelText="Email"
                                name="email"
                                id="email"
                                textInputProps={{
                                    type: 'email',
                                }}
                                value={this.state.email}
                                placeholder="john@doe.com"
                                validationMessage={this.state.errors.email}
                                onChange={(event) =>
                                    this.setState({ email: event.target.value })
                                }
                            />
                            <TextField
                                textInputProps={{
                                    type: 'password',
                                }}
                                labelText="Password"
                                name="password"
                                id="password"
                                className="mt-4"
                                validationMessage={this.state.errors.password}
                                value={this.state.password}
                                onChange={(event) =>
                                    this.setState({
                                        password: event.target.value,
                                    })
                                }
                            />

                            <div className="mt-8 flex justify-between items-center">
                                <CheckboxField
                                    onChange={(event) =>
                                        this.setState({
                                            rememberMe: event.target.checked,
                                        })
                                    }
                                    checked={this.state.rememberMe}
                                    labelText="Remember me"
                                    id="rememberMe"
                                    name="rememberMe"
                                />

                                <Button
                                    type="submit"
                                    buttonType="primary"
                                    disabled={this.state.isLoading}
                                    loading={this.state.isLoading}
                                >
                                    Sign in
                                </Button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        )
    }
}

export default mustBeNotAuthenticated(LoginPage)
