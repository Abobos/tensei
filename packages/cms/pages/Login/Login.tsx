import React, { useState, FormEventHandler } from 'react'
import { TextInput, Checkbox, Button } from '@tensei/components'
import { useHistory, useLocation, Redirect } from 'react-router-dom'

const Login = () => {
  const { push } = useHistory()
  const location = useLocation()
  const [state, setState] = useState<{
    email: string
    password: string
    errors: {
      [key: string]: string
    }
    is_loading: boolean
    is_login: boolean
    submitted?: boolean
  }>({
    email: '',
    password: '',
    errors: {},
    submitted: false,
    is_loading: false,
    is_login: !!location.pathname.match('login')
  })

  const formatErrors = (errors: any[] = []) => {
    let errorsMap: any = {}

    errors.forEach(error => {
      errorsMap[error.field] = error.message
    })

    return errorsMap
  }

  const onSubmit: FormEventHandler<HTMLFormElement> = event => {
    setState({
      ...state,
      errors: {},
      is_loading: true
    })

    event.preventDefault()

    window.Tensei.client
      .post(`auth/login`, {
        email: state.email,
        password: state.password
      })
      .then(() => {
        window.location.href = `/${window.Tensei.state.config.dashboardPath}`

        setState({
          ...state,
          is_loading: false,
          // submitted: true,
          errors: {}
        })
      })
      .catch(({ response }) => {
        setState({
          ...state,
          is_loading: false,
          errors: formatErrors(
            Array.isArray(response.data)
              ? response.data
              : response.data?.error?.errors || []
          )
        })
      })
  }

  return (
    <div className="w-full bg-gray-100 h-screen">
      <div className="max-w-md mx-auto pt-20 md:px-0 px-5">
        <div className="flex justify-center mb-5">
          <img
            className="h-10"
            alt="tensei-logo"
            src="https://res.cloudinary.com/bahdcoder/image/upload/v1604236130/Asset_1_4x_fhcfyg.png"
          />
        </div>
        <div className="border-t-2 border-tensei-primary bg-white shadow-md pt-4 pb-6 px-8">
          <form onSubmit={onSubmit}>
            <TextInput
              id="email"
              name="email"
              label="Email"
              className="mt-4"
              value={state.email}
              placeholder="john@doe.com"
              error={state.errors.email}
              onChange={event =>
                setState({
                  ...state,
                  email: event.target.value
                })
              }
            />
            <TextInput
              id="password"
              name="password"
              label="Password"
              className="mt-4"
              value={state.password}
              placeholder="john@doe.com"
              error={state.errors.password}
              onChange={event =>
                setState({
                  ...state,
                  password: event.target.value
                })
              }
            />
            <div className="flex flex-wrap justify-end items-center mt-2">
              <Button
                primary
                type="submit"
                disabled={state.submitted}
                loading={state.is_loading}
                className="mt-3 md:mt-3 text-center"
              >
                Login
              </Button>
            </div>
          </form>
        </div>
      </div>
    </div>
  )
}

export default Login
