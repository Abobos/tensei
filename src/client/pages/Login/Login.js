import React from 'react'
import Axios from '../../helpers/axios'
import { Button } from 'office-ui-fabric-react/lib/Button'
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox'
import { TextField } from 'office-ui-fabric-react/lib/TextField'

class LoginPage extends React.Component {
  state = {
    username: '',
    password: '',
    rememberMe: false,
  }

  login = (submitEvent) => {
    submitEvent.preventDefault()

    Axios.post('login').then(console.log).catch(console.log)
  }

  render() {
    return (
      <div className="w-full bg-gray-100 h-screen">
        <div className="max-w-md mx-auto pt-20 md:px-0 px-5">
          <div className="flex justify-center mb-5">
            <img
              src="https://strapi.katifrantz.com/admin/3f6f46544e110a51499353fdc9d12bfe.png"
              className="h-10"
              alt="company-logo"
            />
          </div>
          <div className="border-t-2 border-blue-primary bg-white shadow-md py-8 px-8">
            <form onSubmit={this.login} action="">
              <TextField label="Username / Email" placeholder="john@doe.com" />
              <TextField type="password" label="Password" className="mt-4" />

              <div className="mt-8 flex justify-between items-center">
                <Checkbox label="Remember me" />

                <Button type="submit" primary>
                  Sign in
                </Button>
              </div>
            </form>
          </div>
        </div>
      </div>
    )
  }
}

export default LoginPage
