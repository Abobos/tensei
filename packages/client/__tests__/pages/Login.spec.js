import React from 'react'
import { screen, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import '@testing-library/jest-dom'
import { createMemoryHistory } from 'history'
import { MemoryRouter } from 'react-router-dom'

import Auth from '~/store/auth'
import Resources from '~/store/resources'
import Login from '~/pages/Login'

const history = createMemoryHistory()

const LoginSetup = (props) => {
    return (
        <MemoryRouter initialIndex={0} initialEntries={['/auth/login']}>
            <Auth.Provider value={{}}>
                <Resources.Provider
                    value={{
                        resources: [],
                    }}
                >
                    <Login {...props} location={history.location} />
                </Resources.Provider>
            </Auth.Provider>
        </MemoryRouter>
    )
}

describe('Test the Login page', () => {
    beforeEach(() => {
        window.Flamingo = {
            request: {
                post: jest.fn().mockRejectedValue({
                    response: {
                        status: 422,
                        data: {
                            errors: [
                                {
                                    message:
                                        'These credentials do not match our records.',
                                    field: 'email',
                                },
                            ],
                        },
                    },
                }),
            },
        }
    })
    test('should match snapshot', () => {
        window.Flamingo = {
            request: { post: jest.fn(() => Promise.resolve(true)) },
        }
        const { asFragment } = render(<LoginSetup />)

        expect(asFragment()).toMatchSnapshot()
    })
    test('user can login succesfully', () => {
        render(<LoginSetup />)

        const emailField = screen.getByLabelText('Email')
        const passwordField = screen.getByLabelText('password')
        const loginBtn = screen.getByText('Sign in')

        userEvent.type(emailField, 'dozie@gmail.com')
        userEvent.type(passwordField, 'secret')
        userEvent.click(loginBtn)

        expect(window.Flamingo.request.post).toHaveBeenCalled()
    })
    test('user shoule get apporpriate error messages', async () => {
        render(<LoginSetup />)

        const emailField = screen.getByLabelText('Email')
        const passwordField = screen.getByLabelText('password')
        const loginBtn = screen.getByText('Sign in')

        userEvent.type(emailField, 'dozie@emaill.com')
        userEvent.type(passwordField, 'dozie')
        userEvent.click(loginBtn)

        expect(
            await screen.findByText(
                /These credentials do not match our records./i
            )
        ).toBeInTheDocument()
    })
})
