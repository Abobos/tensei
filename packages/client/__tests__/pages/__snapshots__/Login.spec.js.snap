// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Login page should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-gray-100 h-screen"
  >
    <div
      class="max-w-md mx-auto pt-20 md:px-0 px-5"
    >
      <div
        class="flex justify-center mb-5"
      >
        <img
          alt="company-logo"
          class="h-10"
          src="https://strapi.katifrantz.com/admin/3f6f46544e110a51499353fdc9d12bfe.png"
        />
      </div>
      <div
        class="border-t-2 border-blue-primary bg-white shadow-md py-8 px-8"
      >
        <form
          action=""
        >
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK"
            data-test-id="cf-ui-text-field"
            placeholder="john@doe.com"
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="email"
              >
                Email
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="email"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
          </div>
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK mt-4"
            data-test-id="cf-ui-text-field"
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="password"
              >
                Password
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="password"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="password"
                name="password"
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            class="mt-8 flex justify-between items-center"
          >
            <div
              class="ControlledInputField__ControlledInputField___2uIG9"
              data-test-id="cf-ui-checkbox-field"
            >
              <input
                aria-label="Remember me"
                class="ControlledInput__ControlledInput___2XK3j ControlledInputField__ControlledInputField__input___3OMYB"
                data-test-id="cf-ui-controlled-input"
                id="rememberMe"
                name="rememberMe"
                type="checkbox"
                value=""
              />
              <div>
                <label
                  class="FormLabel__FormLabel___3d6zQ ControlledInputField__ControlledInputField__label___a9J52"
                  data-test-id="cf-ui-form-label"
                  for="rememberMe"
                >
                  Remember me
                </label>
              </div>
            </div>
            <button
              class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO"
              data-test-id="cf-ui-button"
              type="submit"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                tabindex="-1"
              >
                <span
                  class="Button__Button__label___3tcOj"
                >
                  Sign in
                </span>
              </span>
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
